### ISSUE TITLE
Some Windows.Services.Store methods need to be called on the UI thread

### ISSUE THREAD

**Issue by fancyydk** [NodeRT/NodeRT#62]
Some methods in the [Windows.Services.Store.StoreContext](https://docs.microsoft.com/en-us/uwp/api/windows.services.store.storecontext) namespace such as

- [RequestDownloadStorePackageUpdatesAsync](https://docs.microsoft.com/en-us/uwp/api/windows.services.store.storecontext#Windows_Services_Store_StoreContext_RequestDownloadStorePackageUpdatesAsync_Windows_Foundation_Collections_IIterable_Windows_Services_Store_StorePackageUpdate__)
- [RequestDownloadAndInstallStorePackageUpdatesAsync](https://docs.microsoft.com/en-us/uwp/api/windows.services.store.storecontext#Windows_Services_Store_StoreContext_RequestDownloadAndInstallStorePackageUpdatesAsync_Windows_Foundation_Collections_IIterable_Windows_Services_Store_StorePackageUpdate__)

must be called on the UI thread. Because NodeRT creates a background task to invoke UWP/WinRT APIs, methods like these return with an error. Having an option to call these APIs directly from the UI thread would be really nice.

**Comment by nadavbar** [comment]
Hi @fancyydk , I'm assuming you are running an electron app using the UWP bridge, right?
While I've experimented with it in the past, I think that the challenging part here might be that you are running from Electron. The way to do such things, in general UWP apps would be to use the core dispatcher and its "RunAsync" method (https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.coredispatcher), what I'm not sure about is if a core dispatcher actually exists in the context of an Electron app which is running from UWP. Do you happen to know how we can check if such thing is doable in the context of the desktop bridge?

**Comment by goldfire** [comment]
@fancyydk @nadavbar Was anything figured out on this front or any workaround found? We've been re-writing our Windows Store app using Electron due to continued issues with the Edge renderer, but we need to handle in-app purchases with `StoreContext.requestPurchaseAsync`, which also requires to be called on the UI thread. Surely there is a way to make this work?

**Comment by goldfire** [comment]
Just as an update, I still haven't found a way to get this to work with `runAsync`, but without documentation it is really difficult to know if I'm just doing this wrong. The following gives the error `Uncaught TypeError: Illegal invocation` on the line calling `runAsync`. Any ideas where this might be going wrong?

```
const Store = require('@nodert-win10-au/windows.services.store');
const Core = require('@nodert-win10-au/windows.applicationmodel.core');
const storeContext = Store.StoreContext.getDefault();

Core.CoreDispatcher.prototype.runAsync(Core.CoreDispatcherPriority.normal, () => {
    storeContext.requestPurchaseAsync(storeId, (result) => {
        // ...
    });
});
```

**Comment by dwcares** [comment]
@goldfire did you ever resolve this?

**Comment by goldfire** [comment]
@dwcares No, we spent quite a bit of time trying to get this to work and eventually gave up on getting this to work. We went back to just using a webpage in a UWP app, but if anyone can get NodeRT working for this we'd obviously prefer to go that route again.

**Comment by elliott-parkinson** [comment]
@goldfire - I'm currently attacking the same issue. But I don't think I've got as far as you have. In order to use the core dispatcher, I need a main window in my application.

I'm relatively new to nodert and have only built c# apps in xaml directly. Did you get around the requirement for a main window, and if so, could you please share your code to create one with me so I can progress?

Below is my current error:

```
Error: A method was called at an unexpected time.

Could not create a new view because the main window has not yet been created
```

**Comment by dwcares** [comment]
for APIs like this i think building a custom native node module is more direct and straightforward. that's what i ended up doing...

**Comment by shanelile** [comment]
@dwcares, do you have any more details on how you got that working from a native node module? When I've tried to call CoreApplication::MainView->CoreWindow->Dispatcher I get a null reference on MainWindow (and had similar issues with the other common calling patterns for getting to the Dispatcher).

**Comment by meteorsnows** [comment]
@dwcares  @bwang29 do you guy have example how to work with the view management api from node.js

**Comment by whitmer** [comment]
Also looking for any working example. I'm trying to add support for the new GazeInput APIs which also need to be called from the UI Thread.

**Comment by tauseefk** [comment]
@whitmer could you try something like https://github.com/Polarrco/windows-IAP-bridge/blob/0cbbe280c3148574cb5ce9bbf8ae4be4b954fd42/src/StoreContext.cpp#L50
You would need to initialize it with a native window handle, from something like Electron or Selenium you can get the native handle and invoke the initialize method.

**Comment by brainexe** [comment]
any other idea here? As I'm fighting with the same topic and try to use the Store in an Electron app. 

A working alternative for the base IAP purchases is https://github.com/Polarrco/windows-IAP-bridge but the method set is quite limited.
